### PR TITLE
Add components that are needed for the OlmoEarth LCC model.

### PR DESCRIPTION
**Pull request by favyen2**
Also update Concatenate so it retains the timestamps so that downstrean we can make use of it (e.g. in the OlmoEarth model when operating over concatenated inputs).

### CODE REVIEW

File: rslearn/models/temporal_transformer.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+import torch
+from einops import rearrange
+from torch import nn
+
+from rslearn.models.component import IntermediateComponent, TokenFeatureMaps
+from rslearn.train.model_context import ModelContext
+
+
+class TemporalTransformer(IntermediateComponent):
```
Comment (gabrieltseng) [inline rslearn/models/temporal_transformer.py:13]: I assume this gets added onto the output of a ViT encoder? I'm curious if this makes a big difference since the whole model is getting finetuned

**Review by gabrieltseng — APPROVED** [review]
It would be nice to add a comment / note describing how these all get wired together in the LCC model

**Comment by favyen2** [comment]
Added notes above about how they are used in LCC. Will add some notes to the code to document how they could be used in general.

File: rslearn/models/temporal_transformer.py (hunk 1 above)
Comment (favyen2) [reply rslearn/models/temporal_transformer.py:13]: added a note that it is most useful for encoder that doesn't reason across timesteps.

For OlmoEarth LCC it is used because it inputs 20 timesteps but that is slow with full attention, so it splits up encoding to two sequences of 10 timesteps, then temporal transformer is used to reason across all 20 timesteps afterwards.

File: rslearn/train/transforms/concatenate.py [hunk 2]
```diff
@@ -34,8 +35,11 @@ def __init__(
                 retain, or empty list to use all bands.
             output_selector: the output selector under which to save the concatenate image.
             concatenate_dim: the dimension against which to concatenate the inputs
+            skip_missing: if True, selectors that are absent from the input/target dicts
+                are silently skipped instead of raising. Useful when concatenating
+                optional inputs (e.g. per-month layers where some months are missing).
```
Comment (yawenzzzz) [inline rslearn/train/transforms/concatenate.py:40]: When we do token pooling, OlmoEarth has the missing token masks that used to masked out those tokens before pooling. I am wondering if we can still have the missing mask here, instead of simply skipping missing tokens, which will make samples temporally misaligned.

Comment (yawenzzzz) [reply rslearn/train/transforms/concatenate.py:40]: https://github.com/allenai/rslearn/blob/778bcdcc7b4933642f3d0e6c8cc377453d656cfd/rslearn/models/olmoearth_pretrain/model.py#L789

File: rslearn/models/temporal_transformer.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    ) -> None:
+        """Create a new TemporalTransformer.
+
+        Args:
+            dim: the token embedding dimension C.
+            depth: number of transformer encoder layers.
+            num_heads: number of attention heads.
+            dim_feedforward: hidden size of the feed-forward network in each layer.
+            dropout: dropout rate in the transformer layers.
+            positional_embedding_num_tokens: if set, add a learned positional
```
Comment (yawenzzzz) [inline rslearn/models/temporal_transformer.py:44]: Same thing here, if the inputs samples are basically temporally misaligned, then will this learned positional embedding still be good for them? also, i guess we can still use the timestamps (like month) based positional encoding here if we carry the timestamps information, and potentially use the `key_padding_mask` to remove the missing timestamps from attention right?

**Review by yawenzzzz — APPROVED** [review]

File: rslearn/models/temporal_transformer.py (hunk 3 above)
Comment (favyen2) [reply rslearn/models/temporal_transformer.py:44]: OK I added mask to TokenFeatureMaps.

**Comment by favyen2** [comment]
@claude can you review this PR?
